### PR TITLE
Add Servo integration via `--browser=servo`

### PR DESCRIPTION
**Pull request by sagudev**
Split off #80 with https://github.com/ErichDonGubler/moz-webgpu-cts/pull/80#issuecomment-1967398153 taken into account. This PR only contains minimal work to acknowledge servo's existence.

It works on servo (if also used with https://github.com/sagudev/moz-webgpu-cts/commit/dddc19ded5ff7b400c4cf4a26989ab6a605b2970), but I do not have m-c clone to test with.

### CODE REVIEW

**Comment by ErichDonGubler** [comment]
@sagudev: Just so you know where I'm at: I'm in favor of this PR, with a few nitpicks that I'm trying to address as `fixup!` changes to facilitate. I'm hoping to get you feedback sometime in the next week; this has been a challenge to get to with the other work I'm doing on Firefox's WebGPU team. 😅

**Comment by sagudev** [comment]
> @sagudev: Just so you know where I'm at: I'm in favor of this PR, with a few nitpicks that I'm trying to address as `fixup!` changes to facilitate. I'm hoping to get you feedback sometime in the next week; this has been a challenge to get to with the other work I'm doing on Firefox's WebGPU team. 😅

No rush, I am currently using #80 myself to update expectations in servo, because it has all I need.

File: moz-webgpu-cts/src/shared.rs
Comment (ErichDonGubler) [inline moz-webgpu-cts/src/shared.rs:492]: issue: We should force calling code to specify what `Browser` they're expecting. This can avoid some issues we otherwise might have with this code falling back to another browser's parent file prefix.

Pushed `fixup!`s 0c54b62, 9ba6deb.

File: moz-webgpu-cts/src/shared.rs
Comment (ErichDonGubler) [inline moz-webgpu-cts/src/shared.rs:1]: nitpick: I don't think the constructor-like methods on `TestScope` carry their weight, esp. given the `visibility` refactors I suggest in other conversations.

Pushed a `fixup!` with 1623e54.

File: moz-webgpu-cts/src/shared.rs
Comment (ErichDonGubler) [reply moz-webgpu-cts/src/shared.rs]: Pushed a `fixup!` with cc57067.

File: moz-webgpu-cts/src/shared.rs
Comment (ErichDonGubler) [inline moz-webgpu-cts/src/shared.rs:897]: issue: This is actually wrong, and it took me a while to figure out why! We weren't preserving the `_webgpu/` prefix on output here, and that's supposed to be here (viz., to mirror how we parse `_webgpu/` in URL path parsing logic).

Pushed a `fixup!` with 9027a2f.

File: moz-webgpu-cts/src/shared.rs
Comment (ErichDonGubler) [inline moz-webgpu-cts/src/shared.rs:675]: nitpick: Let's alphabetize `derive`s; `Copy` comes before `Clone`, but alphabetically comes after.

Pushed as a `fixup!` with df95c12.

File: moz-webgpu-cts/src/main.rs
Comment (ErichDonGubler) [reply moz-webgpu-cts/src/main.rs]: nitpick: Let's use less Firefox-specific language for this, since this is also getting called for Servo.

Pushed as a separate `fix` commit with e86bfd0.

File: moz-webgpu-cts/src/main.rs
Comment (ErichDonGubler) [inline moz-webgpu-cts/src/main.rs:1]: Ah, I just noticed that we have Firefox-specific language in the `UpdateExpected` subcommand. Adjusted as a separate `fix` commit in 132510d.

**Review by ErichDonGubler — APPROVED** [review]
LGTM with some nitpicks and issues that I've resolved with `fixup!` commits. LMK what you think before I merge things; I've tried to leave good notes in the commits themselves, but I've translated feedback to conversations here, too.

File: moz-webgpu-cts/src/shared.rs
Comment (ErichDonGubler) [reply moz-webgpu-cts/src/shared.rs:442]: I have a resolution for this with https://github.com/ErichDonGubler/moz-webgpu-cts/pull/103, which we can save for after this PR.

**Comment by ErichDonGubler** [comment]
Okay! Whew, I think I flushed my queue of feedback here. 😮‍💨 Hope it's not too much!

I actually think that this feedback is sufficient, since I'm punting making things "perfect" to <https://github.com/ErichDonGubler/moz-webgpu-cts/pull/103>. If you think this feedback is fine, then feel free to get history back down to the few commits we intend with `git rebase -i main --autosquash && git push --force-with-lease`. 🙂

---

<sub>I used [Conventional Comments](https://conventionalcomments.org/) in this review! I hope they help with clarity and tone.</sub>

File: moz-webgpu-cts/src/shared.rs
Comment (sagudev) [reply moz-webgpu-cts/src/shared.rs:442]: What does private/public mean in this context? Are WebGPU CTS tests meant to be private or public (I considered them as public), but maybe public is meant for upstream WPT tests?

In servo `_webgpu` is alias to `tests/wpt/webgpu/{meta,tests}` which contains WebGPU CTS. An that's it, those are only WebGPU tests we have.

Comment (ErichDonGubler) [reply moz-webgpu-cts/src/shared.rs:442]: @sagudev: Yep, "public" was supposed to mean "upstream WPT". 😅 I definitely wasn't using the right vocabulary; hopefully the table and the PR clarify the details here.

Comment (sagudev) [reply moz-webgpu-cts/src/shared.rs:442]: Why exactly would upstream path be needed for moz-webgpu-cts? Are there any webgpu test there, or is this some kind of future-proofing for when CTS lands in upstream WPT (or moz-webgpu-cts takes over all expectation updating)?

Comment (ErichDonGubler) [reply moz-webgpu-cts/src/shared.rs:442]: Future-proofing, and also being forced to deal with it early (when I started writing this tool). Due to a fluke in how CI tasks were structured in `mozilla-central`, it was relatively difficult to get reports with only WebGPU tests in WPT test reports. So, I decided to gracefully handle "irrelevant" CTS tests in reports instead.

I don't expect `moz-webgpu-cts` to be useful for general WPT usage. `wpt update` very dynamic in its generation of conditional properties in `wpt update-expectations`, and is configurable at run-time. I don't know who actually uses that amount of power in their daily workflows (CC @jgraham?). However, I suspect that the opinionated, static approach that `moz-webgpu-cts` takes with parsing _and_ properties that get compiled into optimized code (and which allow it to be as fast as it is) may be a significant regression for other users.

Comment (ErichDonGubler) [reply moz-webgpu-cts/src/shared.rs:442]: @sagudev: We could also roll commits from #103 into this PR, if they look good to you.

File: moz-webgpu-cts/src/shared.rs
Comment (ErichDonGubler) [reply moz-webgpu-cts/src/shared.rs:897]: FTR: Finding this actually revealed that we don't test a round-trip between URL path parsing and emission for `TestPath`s, and that made me write <https://github.com/ErichDonGubler/moz-webgpu-cts/pull/101>.

File: moz-webgpu-cts/src/shared.rs
Comment (sagudev) [reply moz-webgpu-cts/src/shared.rs:442]: > Due to a fluke in how CI tasks were structured in mozilla-central, it was relatively difficult to get reports with only WebGPU tests in WPT test reports.

In servo we have special CI option to run only subset of wpt tests (only `_webgpu`). 

> So, I decided to gracefully handle "irrelevant" CTS tests in reports instead.

But "public" results/files are not just skipped? They are also read and formatted, IIRC?

**Comment by sagudev** [comment]
> For better or worse, you are the only person who can actually do this, since the tip branch of this PR isn't part of this repo.

Actually you should be able to to that to, even if the branch is in my fork, because I have `Allowing edits by maintainers` enabled (I think it is by default):

> Allowing edits by maintainers
>
> If checked, users with write access to ErichDonGubler/moz-webgpu-cts can add new commits to your servo-init branch.
>
> You can always change this setting later.

**Comment by ErichDonGubler** [comment]
@sagudev: Ahhh, I see now:

Screenshot 2024-05-11 at 9 35 57 AM

I'm so used to everybody not checking that, thanks!

File: moz-webgpu-cts/src/shared.rs
Comment (ErichDonGubler) [reply moz-webgpu-cts/src/shared.rs:442]: We now have the right filtering in place for Firefox CI. So, we don't generate reports with unrelated entries anymore.  At the time, though, it was easy to work around it with `git add <scope>/meta/webgpu && git restore .` locally. Even now, we don't fully automate the committing of `expected` properties updated via `moz-webgpu-cts`; there's still a human that looks at things that changed. I don't foresee this changing, as we slowly validate what parts of the CTS are stable for us in Firefox.

> But "public" results/files are not just skipped? They are also read and formatted, IIRC?

Sorry, I didn't expect you to be interested in details, so I've been pretty terse. 😅 You're right; report results for any tests outside `<scope>/meta/webgpu` will still get processed and written out by `moz-webgpu-cts` as with tests it's intended to be used for.

Comment (sagudev) [reply moz-webgpu-cts/src/shared.rs:442]: > report results for any tests outside <scope>/meta/webgpu will still get processed and written out by moz-webgpu-cts as with tests it's intended to be used for.

That's what's been confusing me the most. In servo there are a lot of "invalid" and not formatted expectations files, so that's why we cannot have `moz-webgpu-cts` try to load them.

**Comment by ErichDonGubler** [comment]
@sagudev: FYI: Just piggybacked a fix to CI with [062b2b9](https://github.com/ErichDonGubler/moz-webgpu-cts/pull/92/commits/062b2b95884e9c7deb6d321a1ede02850829416e), so this could auto-merge. 🙂

File: moz-webgpu-cts/src/shared.rs
Comment (ErichDonGubler) [reply moz-webgpu-cts/src/shared.rs:442]: Is the Servo project actually interested in using `moz-webgpu-cts` in a broader scope? Or is that merely a pain point that's forced you to narrow what metadata `moz-webgpu-cts` is incidentally exposed to? 🤔

Comment (sagudev) [reply moz-webgpu-cts/src/shared.rs:442]: > Is the Servo project actually interested in using `moz-webgpu-cts` in a broader scope?

Formatting would be useful as I do not think any exists for expectations (or at least we aren't using any), but there is currently nothing planned.

> Or is that merely a pain point that's forced you to narrow what metadata `moz-webgpu-cts` is incidentally exposed to? 🤔

This is exactly what happened.

Comment (ErichDonGubler) [reply moz-webgpu-cts/src/shared.rs:442]: > Formatting would be useful as I do not think any exists for expectations (or at least we aren't using any), but there is currently nothing planned.

I suspect that a general-purpose formatter would be very difficult to get consensus on, as opposed to the very opinionated approach here (CC @jgraham). I don't see any upstream issues in [`wpt`](https://github.com/web-platform-tests/wpt) or [`rfcs`](https://github.com/web-platform-tests/rfcs) for it. However, you _could_ use `whippit` now to create your own parser, and then a custom emitter like [`format_file`](https://github.com/ErichDonGubler/moz-webgpu-cts/blob/6b4d64591b5f54f7f6e4d86c389edf88165e7324/moz-webgpu-cts/src/metadata.rs#L694) to enforce things like property order and spacing. AIUI that's really the full breadth of things you can control, with the metadata language.

**Pull request merged**
